x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/chunking/SentenceBoundaryChunker.java

```diff
@@ -75,24 +76,46 @@ public List<String> chunk(String input, int maxNumberWordsPerChunk) {
         int sentenceStart = 0;
         int chunkWordCount = 0;
 
+        int wordsInPrecedingSentenceCount = 0;
+        int previousSentenceStart = 0;
+
         int boundary = sentenceIterator.next();
 
         while (boundary != BreakIterator.DONE) {
             int sentenceEnd = sentenceIterator.current();
-            int countWordsInSentence = countWords(sentenceStart, sentenceEnd);
+            int wordsInSentenceCount = countWords(sentenceStart, sentenceEnd);
 
-            if (chunkWordCount + countWordsInSentence > maxNumberWordsPerChunk) {
+            if (chunkWordCount + wordsInSentenceCount > maxNumberWordsPerChunk) {
                 // over the max chunk size, roll back to the last sentence
 
+                int nextChunkWordCount = wordsInSentenceCount;
                 if (chunkWordCount > 0) {
                     // add a new chunk containing all the input up to this sentence
                     chunks.add(input.substring(chunkStart, chunkEnd));
-                    chunkStart = chunkEnd;
-                    chunkWordCount = countWordsInSentence; // the next chunk will contain this sentence
+
+                    if (includePrecedingSentence) {
+                        if (wordsInPrecedingSentenceCount + wordsInSentenceCount > maxNumberWordsPerChunk) {
+                            // cut the last sentence
+                            int numWordsToSkip = numWordsToSkipInPreviousSentence(wordsInPrecedingSentenceCount, maxNumberWordsPerChunk);
+
+                            chunkStart = skipWords(input, previousSentenceStart, numWordsToSkip);
+                            chunkWordCount = (wordsInPrecedingSentenceCount - numWordsToSkip) + wordsInSentenceCount;
+                        } else {
+                            chunkWordCount = wordsInPrecedingSentenceCount + wordsInSentenceCount;
+                            chunkStart = previousSentenceStart;
+                        }
+
+                        nextChunkWordCount = chunkWordCount;
+                    } else {
+                        chunkStart = chunkEnd;
+                        chunkWordCount = wordsInSentenceCount; // the next chunk will contain this sentence
+                    }
```

Comment:
nit: I feel like this should be a standalone function. Something like `startNextChunk`. Personally, I think it would make this code easier to follow.